### PR TITLE
Fix tsdb codec when doc-values spread in two blocks

### PR DESCRIPTION
**Pull request by dnhatn**
Currently, loading ordinals multiple times (after advanceExact) for documents with values spread across multiple blocks in the TSDB codec will fail due to the absence of re-seeking for the ordinals block. Doc-values of a document can spread across multiple blocks in two cases: when it has more than 128 values or when it exceeds the remaining space in the current block.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/codec/tsdb/ES87TSDBDocValuesProducer.java [hunk 1]
```diff
@@ -1106,8 +1106,8 @@ long advance(long index) throws IOException {
                 final long blockIndex = index >>> ES87TSDBDocValuesFormat.NUMERIC_BLOCK_SHIFT;
                 final int blockInIndex = (int) (index & ES87TSDBDocValuesFormat.NUMERIC_BLOCK_MASK);
                 if (blockIndex != currentBlockIndex) {
-                    assert blockIndex > currentBlockIndex;
-                    if (blockIndex - 1 > currentBlockIndex) {
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/codec/tsdb/ES87TSDBDocValuesProducer.java:1110]: I see this exact pattern in `NumericDocValues#longValue`. Do we have off-by-one errors there as well?

https://github.com/elastic/elasticsearch/blob/ae69a61fdb326e69e422cc4e055622d9c52bcf36/server/src/main/java/org/elasticsearch/index/codec/tsdb/ES87TSDBDocValuesProducer.java#L1015

https://github.com/elastic/elasticsearch/blob/ae69a61fdb326e69e422cc4e055622d9c52bcf36/server/src/main/java/org/elasticsearch/index/codec/tsdb/ES87TSDBDocValuesProducer.java#L1075

?

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/codec/tsdb/ES87TSDBDocValuesProducer.java:1110]: Thanks Bent. These are single value paths, so they shouldn't be affected by this bug. Anyway, I've extended the tests and made the seeking patterns more consistent in this class.

**Review by benwtrent — APPROVED** [review]

**Comment by dnhatn** [comment]
@benwtrent Thanks for review.

**Review by martijnvg — COMMENTED** [review]
LGTM2

**Comment by jpountz** [comment]
Thanks for fixing! I tried to understand why `BaseDocValuesFormatTestCase#testSortedSetVariableLengthVsStoredFields` did not catch this bug, and this seems to be because it only uses `nextDoc()`, not `advance()`/`advanceExact()`? Could you improve `BaseDocValuesFormatTestCase#doTestSortedSetVsStoredFields` in Lucene to add more coverage for this?

**Comment by dnhatn** [comment]
@jpountz +1, I will open a PR in the Lucene repository.

**Comment by dnhatn** [comment]
I've opened https://github.com/apache/lucene/pull/13346

**Pull request merged**